blocks_vertical/procedures.js

```diff
@@ -29,6 +29,281 @@ goog.require('Blockly.Blocks');
+  // Split the proc into components, by %n, %b, and %s (ignoring escaped).
+  var procComponents = this.procCode_.split(/(?=[^\\]\%[nbs])/);
+  procComponents = procComponents.map(function(c) {
+    return c.trim(); // Strip whitespace.
+  });
+  // Create inputs and shadow blocks as appropriate.
+  var inputPrefix = 'input';
+  var inputCount = 0;
+  for (var i = 0, component; component = procComponents[i]; i++) {
+    var newLabel;
+    if (component.substring(0, 1) == '%') {
+      var inputType = component.substring(1, 2);
+      if (!(inputType == 'n' || inputType == 'b' || inputType == 's')) {
+        throw new Error(
+            'Found an custom procedure with an invalid type: ' + inputType);
+      }
+
+      newLabel = component.substring(2).trim();
+
+      var id = this.argumentIds_[inputCount];
+      var oldBlock = null;
+      if (connectionMap && (id in connectionMap)) {
+        oldBlock = connectionMap[id];
+      }
+
+      var inputName = inputPrefix + (inputCount++);
+      var input = this.createInput_(inputType, inputName, oldBlock, id,
+          connectionMap);
+      params[id] = input;
+    } else {
+      newLabel = component.trim();
+    }
+    this.appendDummyInput().appendField(newLabel.replace(/\\%/, '%'));
+  }
+  return params;
+};
+
+/**
+ * Build a DOM node representing a shadow block of the given type.
+ * @param {string} type One of 's' (string) or 'n' (number).
+ * @return {!Element} The DOM node representing the new shadow block.
+ * @private
+ * @this Blockly.Block
+ */
+Blockly.ScratchBlocks.ProcedureUtils.buildShadowDom_ = function(type) {
+  var xmlStart = '<xml xmlns="http://www.w3.org/1999/xhtml">';
+  var xmlEnd = '</xml>';
+  if (type == 'n') {
+    var shadow = '<shadow type="math_number">' +
+        '<field name="NUM">10</field>' +
+        '</shadow>';
+  } else {
+    var shadow = '<shadow type="text"></shadow>';
+  }
+  return Blockly.Xml.textToDom(xmlStart + shadow + xmlEnd).firstChild;
+};
+
+/**
+ * Reattach a block to the correct input after a mutation, and give the
+ * connection a shadow DOM based on its input type.
+ * @param {!Blockly.Input} input The value input to attach a block to.
+ * @param {string} inputType One of 'b' (boolean), 's' (string) or 'n' (number).
+ * @param {Blockly.Block} oldBlock The block that needs to be attached to the
+ *     input, or null if none existed.
+ * @param {string} id The ID of the current parameter.
+ * @param {!Object.<string, Blockly.Block>} connectionMap An object mapping
+ *     parameter IDs to the blocks that were connected to those IDs at the
+ *     beginning of the mutation.
+ * @private
+ * @this Blockly.Block
+ */
+Blockly.ScratchBlocks.ProcedureUtils.reattachBlock_ = function(input, inputType,
+    oldBlock, id, connectionMap) {
+  if (inputType == 'n' || inputType == 's') {
+    input.connection.setShadowDom(this.buildShadowDom_(inputType));
+  } else {
+    input.setCheck('Boolean');
+  }
+  connectionMap[id] = null;
```

Comment:
hmm, I don't understand why the entry in the connection map gets set to null here. Which means I'm probably misunderstanding something higher level.  The connectionMap is mutable during the mutation? I thought it was supposed to be the state of the block before the mutation was applied?

Reply:
I may be overloading connectionMap.  After reattaching all blocks that I can, I want to delete all shadow blocks that were not reattached--and probably bump all non-shadow blocks that were not reattached.

The connection map has all of the connected blocks from right before the mutation was applied.  I'm using it to figure out if I have a block to reconnect to a given input.  Each block can only be reattached once so it's safe to remove it here.  After creating all of the inputs, any remaining blocks in the map must need to be deleted or bumped.

The alternative is to create a separate list of blocks (orphanBlocks?), add all blocks to it at the beginning, and then remove blocks from it as I go.  I'm happy to do that if it's clearer.